dhis2/dhis2-core
# feat: API tokens version 2

## Summary

1. Refactoring of utility classes and methods related to API tokens.
2. New PAT token format, using longer tokens (256-bit) and a base-62 CRC32 checksum. PATs are now also upgradeable as a result of the refactoring. [DHIS2-15338](https://dhis2.atlassian.net/browse/DHIS2-15338)
3. Add authentication details with IP address in the authentication log [DHIS2-13838](https://dhis2.atlassian.net/browse/DHIS2-13838)
4. Remove token from authentication cache when a token is deleted. [DHIS2-15339](https://dhis2.atlassian.net/browse/DHIS2-15339)

### Automatic tests
Everything should be covered by automatic tests.

### Manual tests

2.  Checkout latest master and start the server and create a PAT, this will be a version 1 token.
  * Checkout this branch and do the same, this will be a token with version 2.
  * Now try to authenticate with both.
3. Do an API request with a PAT and observe the log now logs the IP address with the authentication.
4. Try to delete a token that you already used, when using it again, it should not work.      



[DHIS2-13838]: https://dhis2.atlassian.net/browse/DHIS2-13838?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DHIS2-15338]: https://dhis2.atlassian.net/browse/DHIS2-15338?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DHIS2-15339]: https://dhis2.atlassian.net/browse/DHIS2-15339?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
